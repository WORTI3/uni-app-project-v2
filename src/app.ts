// Dependencies
import express, { ErrorRequestHandler } from 'express';
import nunjucks from 'nunjucks';
import createError from 'http-errors';
import path from 'path';
import cookieParser from 'cookie-parser';
import session from 'express-session';
import csrf from 'csurf';
import sqlite3 from 'sqlite3';
import sqliteStoreFactory from 'express-session-sqlite';
// App routers
import { initPassport } from './middleware/passport';
import { homeRouter } from './routes/home';
import { authRouter } from './routes/auth';
import { indexRouter } from './routes/index';
import { Session } from './types';

const app = express();
// Db
const SqliteStore = sqliteStoreFactory(session);
const store = new SqliteStore({
  driver: sqlite3.Database,
  path: './src/database/sessions.db', // Specify the path to the SQLite database file
  ttl: 86400, // Session Time To Live in seconds
  cleanupInterval: 300000, // Cleanup interval in milliseconds for deleting expired sessions
});

// Nunjucks configuration
nunjucks.configure(['./src/views', './src/_layouts'], {
  autoescape: true,
  express: app,
});
// View engine setup
app.set('view engine', 'njk');
app.use(express.json());
app.use(express.urlencoded({ extended: false }));
app.use(cookieParser());

// Static assets (main.css)
app.use(express.static(path.join('./', 'public')));

// Adds CSRF protection to the application.
app.use(csrf({ cookie: true }));

// Init session
app.use(
  session({
    secret: 'nobody knows',
    resave: false,
    saveUninitialized: false,
    store,
  }),
);
// passport middleware
initPassport(app);

/**
 * Sets messages and tone to locals from the current session.
 * @param {Object} req - The request object.
 * @param {Object} res - The response object.
 * @param {Function} next - The next middleware function.
 */
app.use((req, res, next) => {
  const session = req.session as Session;

  if (!session.errorFields) {
    // when no session is present
    const inputValues = {
      username: req.body.username ?? '',
      password: '', // dont want to pass password value through here
    };
    const fields = [];
    for (const [key, value] of Object.entries(inputValues)) {
      fields.push({
        field: key,
        value: value,
      });
    }
    session.errorFields = fields;
  }

  // todo remove messages and use errorFields.error

  if (session.messages || session.errorFields) {
    console.log('error fields:', session.errorFields);

    const msgs = session?.messages ?? [];
    const tone = session.msgTone ?? null;
    res.locals.messages = msgs;
    res.locals.msgTone = tone;
    res.locals.hasMessages =
      Boolean(msgs.length) || Boolean(session.errorFields);
    res.locals.errorFields = session.errorFields;
    session.messages = [];
    // session.errorFields = [];
    session.msgTone = undefined;
  }
  next();
});

/**
 * Middleware function that adds a CSRF token to the response locals object.
 * If the environment is set to "test", a hardcoded token is used instead.
 * @param {Object} req - The request object.
 * @param {Object} res - The response object.
 * @param {Function} next - The next middleware function.
 * @returns None
 */
app.use((req, res, next) => {
  res.locals.csrfToken = req.csrfToken();
  next();
});

// App routes
app.use('/', homeRouter);
app.use('/', indexRouter);
app.use('/', authRouter);

// Catch 404 and forward to error handler
app.use((_req, _res, next) => {
  next(createError(404));
});

// Error handler
const errorHandler: ErrorRequestHandler = (err, req, res, _next) => {
  // Set locals, only providing error in development
  res.locals.message = err.message;
  res.locals.error = req.app.get('env') === 'development' ? err : {};
  res.status(err.status || 500).render('error');
};

app.use(errorHandler);

export default app;
