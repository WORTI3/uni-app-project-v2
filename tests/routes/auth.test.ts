import request from "supertest";
import cheerio from "cheerio";
import app from "../../src/app";

describe("GET /signup", () => {
  it("should return 200 status code match snapshot", async () => {
    // given / when
    const result = await request(app).get("/signup").expect(200);
    
    // then
    expect(result.get("Content-Type")).toContain("text/html");
    // Remove CSRF token from the response text for snapshot matching
    const resultTextNoCSRF = result.text.replace(/<input type="hidden" name="_csrf" value=".*">/g, '');
    expect(resultTextNoCSRF).toMatchSnapshot();
  });

  it("should render add content", async () => {
    // given / when
    const result = await request(app).get("/signup").expect(200);

    // then
    const $ = cheerio.load(result.text);
    const para = $("p").text();
    expect(para).toBe("Already have an account? Sign inPassword must be at least 5 characters containing 1 uppercase letter, 1 lowercase letter, 1 number and 1 special character");
    expect($("a").length).toBe(4);
  });
});

describe("GET /login", () => {
  it("should return 200 status code match snapshot", async () => {
    // given / when
    const result = await request(app).get("/login").expect(200);

    // then
    expect(result.get("Content-Type")).toContain("text/html");
    const resultTextNoCSRF = result.text.replace(/<input type="hidden" name="_csrf" value=".*">/g, '');
    expect(resultTextNoCSRF).toMatchSnapshot();
  });

  it("should render add content", async () => {
    // given / when
    const result = await request(app).get("/login").expect(200);

    // then
    const $ = cheerio.load(result.text);
    const para = $("p").text();
    expect(para).toBe("Don't have an account? Sign Up");
    expect($("a").length).toBe(4);
  });
});
