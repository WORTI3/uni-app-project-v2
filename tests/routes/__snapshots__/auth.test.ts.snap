// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GET /login should return 200 status code match snapshot 1`] = `
"
<!DOCTYPE html>
<html lang="en">
    
	<head>
		<meta charset="utf-8">
		<meta name="viewport" content="width=device-width, initial-scale=1">
        <meta name="description" content="Faultifier - a ticket handling admin service.">
        <meta name="theme-color" content="blue">
		<title>Login - Faultifier</title>
		<link rel="stylesheet" type="text/css" href="/main.css">
	</head>

    <div class="flex flex-col h-screen justify-between overflow-hidden">
        <nav class="flex items-center justify-between flex-wrap bg-gray-100 p-6">
    <div class="flex items-center flex-shrink-0 text-picton-blue-600 mr-6">
        <a href="/"><span class="font-bold text-xl tracking-tight font-sans">FAULTIFIER</span></a>
    </div>
    
    
</nav>

        <body class="bg-gradient-to-tr from-picton-blue-600 to-yellow-500 overflow-hidden">
            <main>
<div class="p-12 bg-white sm:mx-auto rounded-md w-1/2 max-sm:w-full max-sm:p-6">
	<a href="/" class="float-left text-sm font-bold text-sky-600 hover:text-sky-600">Go back</a><br>
	
	
    <div class="mb-5">
		<h3 class="font-semibold text-2xl text-gray-800">Sign In</h3>
		<p class="text-gray-400">Don&#39;t have an account? <a href="/signup" class="text-sm font-bold text-sky-600 hover:text-sky-600">Sign Up</a></p>
	</div>

	
	

  

  

	
	<form action="/login/password" method="post" novalidate>
			
    	
			
			

  
  
  
  
  
  <div class="relative">
    <div class="flex flex-row items-center border-b sm:border-b-0 w-full sm:w-auto py-2 sm:pb-0">
      
      <label for='username' class="text-gray-500 ">Username or Email</label>
    </div>
    <input id="username" name="username" type="text" value="" placeholder="username" autocomplete="username" required  class="w-full text-sm px-4 py-3 bg-gray-200 focus:bg-gray-100 border border-gray-200 rounded-lg focus:outline-none focus:border-picton-blue-500 mb-6">  
  </div>

  
  
  
  
  
  <div class="relative">
    <div class="flex flex-row items-center border-b sm:border-b-0 w-full sm:w-auto py-2 sm:pb-0">
      
      <label for='password' class="text-gray-500 ">Password</label>
    </div>
    <input id="password" name="password" type="password" value="" placeholder="password" autocomplete="password" required  class="w-full text-sm px-4 py-3 bg-gray-200 focus:bg-gray-100 border border-gray-200 rounded-lg focus:outline-none focus:border-picton-blue-500 mb-6">  
  </div>


			
    <button type="submit" class="w-full flex justify-center bg-picton-blue-500 hover:bg-picton-blue-700 text-gray-100 p-3  rounded-lg tracking-wide font-semibold  cursor-pointer transition ease-in duration-500">
        Sign in
    </button>

	</form>
</div>
</main>
        </body>

        
        <footer class="lg:text-left bg-gray-100 text-gray-600">
            <div class="text-center p-6">
                <span>© 2023 </span>
                <a class="text-gray-600 font-semibold" href="https://github.com/WORTI3">Wortie</a>
            </div>
        </footer>
    </div>
</html>"
`;

exports[`GET /signup should return 200 status code match snapshot 1`] = `
"
<!DOCTYPE html>
<html lang="en">
    
	<head>
		<meta charset="utf-8">
		<meta name="viewport" content="width=device-width, initial-scale=1">
        <meta name="description" content="Faultifier - a ticket handling admin service.">
        <meta name="theme-color" content="blue">
		<title>Sign up - Faultifier</title>
		<link rel="stylesheet" type="text/css" href="/main.css">
	</head>

    <div class="flex flex-col h-screen justify-between overflow-hidden">
        <nav class="flex items-center justify-between flex-wrap bg-gray-100 p-6">
    <div class="flex items-center flex-shrink-0 text-picton-blue-600 mr-6">
        <a href="/"><span class="font-bold text-xl tracking-tight font-sans">FAULTIFIER</span></a>
    </div>
    
    
</nav>

        <body class="bg-gradient-to-tr from-picton-blue-600 to-yellow-500 overflow-hidden">
            <main>
<div class="p-12 bg-white sm:mx-auto rounded-md w-1/2 max-sm:w-full max-sm:p-6">
	<a href="/" class="float-left text-sm font-bold text-sky-600 hover:text-sky-600">Go back</a><br>
	
    <div class="mb-5">
		<h3 class="font-semibold text-2xl text-gray-800">Sign up</h3>
		<p class="text-gray-400">Already have an account? <a href="/login" class="text-sm font-bold text-sky-600 hover:text-sky-600">Sign in</a></p>
	</div>

	
	

  

  

	<form action="/signup" method="post" novalidate>
    	
    	
			
			

  
  
  
  
  
  <div class="relative">
    <div class="flex flex-row items-center border-b sm:border-b-0 w-full sm:w-auto py-2 sm:pb-0">
      
      <label for='username' class="text-gray-500 ">Username or Email</label>
    </div>
    <input id="username" name="username" type="text" value="" placeholder="username" autocomplete="username" required  class="w-full text-sm px-4 py-3 bg-gray-200 focus:bg-gray-100 border border-gray-200 rounded-lg focus:outline-none focus:border-picton-blue-500 mb-6">  
  </div>

  
  
  
  
  
  <div class="relative">
    <div class="flex flex-row items-center border-b sm:border-b-0 w-full sm:w-auto py-2 sm:pb-0">
      
      <label for='password' class="text-gray-500 ">Password</label>
    </div>
    <input id="password" name="password" type="password" value="" placeholder="password" autocomplete="password" required  class="w-full text-sm px-4 py-3 bg-gray-200 focus:bg-gray-100 border border-gray-200 rounded-lg focus:outline-none focus:border-picton-blue-500 mb-6">  
  </div>


			<p class="-mt-6 mb-4 font-semibold text-xs mx-2 text-gray-500">Password must be at least 5 characters containing 1 uppercase letter, 1 lowercase letter, 1 number and 1 special character</p>
			
    <button type="submit" class="w-full flex justify-center bg-picton-blue-500 hover:bg-picton-blue-700 text-gray-100 p-3  rounded-lg tracking-wide font-semibold  cursor-pointer transition ease-in duration-500">
        Sign up
    </button>

	</form>
</div>
</main>
        </body>

        
        <footer class="lg:text-left bg-gray-100 text-gray-600">
            <div class="text-center p-6">
                <span>© 2023 </span>
                <a class="text-gray-600 font-semibold" href="https://github.com/WORTI3">Wortie</a>
            </div>
        </footer>
    </div>
</html>"
`;
